### ISSUE TITLE
OpenTracing duplicates user created spans during replay

### ISSUE THREAD

**Issue by Spikhalskiy** [temporalio/sdk-java#952]
## Expected Behavior

If we have the following code somewhere in workflow:

```
void execute() {
  ...
  withSpan(() -> {
    activity.runActivity();
  })
  ...
  //end of replay somewhere down from the above code block
}
```

If we replay the code above, we should ignore the span creation, because this span was already reported during the original execution.

## Actual Behavior

The second span will be created during replay.

## Proposed solution

We should create a thin layer on top of `Tracer`, `ScopeManager`, `Scope` or `Span` that will check if we are still replaying on span finalization and if we are - discard the span instead of finalizing it.

## Notes
The issue is raised here:
#951
https://github.com/temporalio/sdk-java/pull/951#pullrequestreview-844189585

And the solution is originally proposed here:
https://github.com/temporalio/sdk-java/pull/951#issuecomment-1005896960

**Comment by bergundy** [comment]
> But what is true, the user-created span may be ignored during replay (depends on its location) with my original proposal. And in your sample, the activity may end up directly under the WorkflowRun span. But this is a smaller issue, let me think about it.
I really believe that we can resolve it just as inside a thin Tracer wrapper that can make some smart "buffering" of spans during replays inside and track when isReplaying switches from true to false.
Or even simpler, the tracer can always open spans when requested but check isReplaying on .finish() call and discard the span if isReplaying is still true. At first sight, it looks like it will work.

So the problem is that right now with a user created span the activity span will be parentless because the parent span (the user created one) is never emitted.
The only way to fix this is to recreate the same span with the exact same spanId on replay.

**Comment by Spikhalskiy** [comment]
@bergundy it's not. For this specific reason, JavaSDK always creates WorkflowRun span. As you can see in the test that I pointed: https://github.com/temporalio/sdk-java/blob/a2bc9373201dd58a607197ac39b7f87c1c0059c2/temporal-opentracing/src/test/java/io/temporal/opentracing/ActivityFailureTest.java#L131. So, the activity span will not be parentless, it will have WorkflowRun span as a parent.

But yes, it will not have a user-created span as a parent. I'm fairly confident that the proposed solution will address it well.

**Comment by Spikhalskiy** [comment]
@bergundy 

> The only way to fix this is to recreate the same span with the exact same spanId on replay.

It's not. Even more, OpenTracing and OpenTelemetry specification doesn't allow this (creation of a span with a specific spanId) explicitly.

Each run has its own WorkflowRun span. The user span should be duplicated under each WorkflowRun span and contain only activities that were executed and not replayed in that run. This will be a solution congruent with the current JavaSDK approach and OpenTracing/OpenTelemetry spec.

**Comment by bergundy** [comment]
> A span that was finished can't be just reopened for appending
In the case I've illustrated in #951 the span is unfinished (the workflow is evicted from cache before the activity can complete)

**Comment by Spikhalskiy** [comment]
> In the case I've illustrated in #951 the span is unfinished (the workflow is evicted from cache before the activity can complete)

Well, maybe in TS, not in Java.
In Java the [conventional code](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/ScopeManager.java#L45) working with spans look like this:

```
           Span span = tracer.buildSpan("...").start();
           try (Scope scope = tracer.scopeManager().activate(span)) {
               span.setTag("...", "...");
               ...
           } catch (Exception e) {
               span.log(...);
           } finally {
               span.finish();
           }
```

Scope and span will be finished and reported in case of the eviction.

**Comment by bergundy** [comment]
Okay, I see, it works differently than in TS where we completely discard the workflow from memory when it's evicted without running any cleanup.

So in this case if the workflow is replayed before it can execute activity2, activity2's parent span will be the root workflow span which isn't that bad.

**Comment by Spikhalskiy** [comment]
@bergundy Yeah, in Java you can't just nuke something without running the corresponding `.finally()` blocks.
Which has its own advantages (here) and creates its own challenges for us (potential non-determinism if somebody puts an activity execution in `.finally()`).

**Comment by mfateev** [comment]
I don't think that approach that doesn't create a span on replay would work. The problem is that a newly executed code can create a child span and needs a parent span. But parent span wouldn't be present on replay. The ideal solution would allow assigning span-id deterministically on replay. But unfortunately none of the tracing libraries supports assigning span ids.

The other solution is to serialize the span into the history and then recreate it from the serialized data on replay.

**Comment by Spikhalskiy** [comment]
> But parent span wouldn't be present on replay.
It will.

> The ideal solution would allow assigning span-id deterministically on replay.
Against the spec, as pointed by me above.
